### PR TITLE
chore: use Rslib to bundle `@rspack/core`

### PR DESCRIPTION
**Pull request by Timeless0911**
## Summary

This PR use Rslib to bundle `@rspack/core`.

**Bundle size:**

- **tsup**:
	- `dist/cssExtractHmr.js`: 5.70 KB = 5.57 kB
	- `dist/cssExtractLoader.js`: 10.27 KB = 10.03 kB
	- `dist/worker.js`: 66.83 KB = 65.26 kB
	- `dist/index.js`: 695.87 KB = 712.57 kB
- **Rslib**:
 	- `dist/cssExtractHmr.js`: 6.2 kB
	- `dist/cssExtractLoader.js`: 10.6 kB
	- `dist/worker.js`: 71.5 kB
	- `dist/index.js`: 772.0 kB

Below are some things that need to be explained and documented:

- Remove `tsc` since Rslib use Typescript Compiler API to genenrate declaration files.
- Remove `tsc-alias` since Rslib internally support redirect alias in declaration files via `@ast-grep/napi`.
- Add a new `tsconfig.api.extractor.json` file to override `paths` in `tsconfig.json` since api-extractor will throw errors if correct paths which in prebundle folder `compiled` are applied.
image
- Add an empty `CachedInputFileSystem.d.ts` file to prevent ts error when prebundle.
- Remove the modification for prebundled `index.d.ts` for `ResolveRequest` type export since it is not used now.
- Use `import z from "zod";` instead of `import { z } from "zod";`
	- Now: image
	- Before: image
- Remove `module.exports = rspack;` in source code and add a footer `js: "module.exports = __webpack_exports__.default;"` instead to make default export in cjs work.
- Add a footer `module.exports = __webpack_exports__.default;` when bundle `worker.js` to make default export in cjs work.
- Remove some `@ts-expect-error` since we use correct `paths` in `tsconfig.json` now.  
- Modify `{ __webpack_require__ }` to `{ __webpack_require__: __webpack_require__ }` to avoid parser bug. can be reset when https://github.com/web-infra-dev/rspack/pull/10036 is released in Rslib.

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/rspack-test-tools/tests/hotCases/css/recovery/__snapshots__/web/2.snap.txt [hunk 1]
```diff
@@ -37,228 +37,192 @@ self["webpackHotUpdate"]('main', {
 /*!***************************************************!*\
   !*** ../../../../../rspack/dist/cssExtractHmr.js ***!
   \***************************************************/
-(function (module) {
+(function (__unused_webpack_module, exports) {
```
Comment (Timeless0911) [inline packages/rspack-test-tools/tests/hotCases/css/recovery/__snapshots__/web/2.snap.txt:40]: https://github.com/web-infra-dev/rspack/pull/10016

File: packages/rspack/src/taps/compilation.ts [hunk 2]
```diff
@@ -184,7 +184,7 @@ export const createCompilationHooksRegisters: CreatePartialRegisters<
 											codeGenerationResult: new CodeGenerationResult(result),
 											moduleObject
 										},
-										{ __webpack_require__ }
+										{ __webpack_require__: __webpack_require__ }
```
[the remark this answers is not in the document]
Comment (Timeless0911) [reply packages/rspack/src/taps/compilation.ts:187]: https://github.com/web-infra-dev/rspack/pull/10036

**Comment by fi3ework** [comment]
I think the bloated bundled size is due to Rspack doesn't allow concatenated module for CJS output?

**Review by chenjiahan — APPROVED** [review]
LGTM👍 

After checking the snapshots, I found that the `__nested_webpack_require_18_37__` variable is quite ugly.

Maybe we should use `scripts/prepare-container-runtime.js` (or some other methods) to transform `src/runtime/cssExtractHmr.ts` to improve the generated code.

**Comment by chenjiahan** [comment]
> I think the bloated bundled size is due to Rspack doesn't allow concatenated module for CJS output?

We can use the same minify config as Rsbuild to output a smaller bundle size. https://github.com/web-infra-dev/rsbuild/blob/6c2a605b7cfc2c07a8d1f47ed706d6abe6ab12c3/scripts/config/rslib.config.ts#L9-L20

**Comment by chenjiahan** [comment]
Let's merge this PR first and further optimize the bundle size and output code readability in subsequent PRs.

**Comment by Timeless0911** [comment]
> I think the bloated bundled size is due to Rspack doesn't allow concatenated module for CJS output?

@fi3ework In fact, if we set `optimization.concatenateModules: false`, the `index.js` is 1013.6 kB, which is much larger than 772.2 kB when `concatenateModules` is not turned on.

**Comment by Timeless0911** [comment]
`After checking the snapshots, I found that the __nested_webpack_require_18_37__ variable is quite ugly.`

The output of `src/runtime/cssExtractHmr.ts` is correct with `__webpack_require__`.
image

@chenjiahan The snapshots seems to show the result of `cssExtractHmr.js` file being re-bundled by Rspack which is renamed to prevent reference to the same runtime. 

This renaming behavior found some [edge cases](https://github.com/web-infra-dev/rspack/pull/10016), but it should be in line with expectations in this test case and needs to be confirmed again. @JSerFeng cc

**Comment by chenjiahan** [comment]
I'm not saying Rslib is built incorrectly, just that it's not easy to read.

In an Rspack application, `cssExtractHmr.js` is always re-bundled, so the generated code that users see will be `__nested_webpack_require_18_37__`

**Comment by Timeless0911** [comment]
> Maybe we should use scripts/prepare-container-runtime.js (or some other methods) to transform src/runtime/cssExtractHmr.ts to improve the generated code.

We can transform via [SWC Compiler API](https://github.com/web-infra-dev/rspack/pull/10017) provided by Rspack in the near future.
